pallets/subtensor/tests/math.rs

```diff
@@ -0,0 +1,2344 @@
+    let mut x1: Vec<I32F32> = vec![
+        I32F32::from_num(1.0),
+        I32F32::from_num(10.0),
+        I32F32::from_num(30.0),
+    ];
+    inplace_normalize(&mut x1);
+    assert_vec_compare(
+        &x1,
+        &[
+            I32F32::from_num(0.0243902437),
+            I32F32::from_num(0.243902439),
+            I32F32::from_num(0.7317073171),
+        ],
+        epsilon,
+    );
+    let mut x2: Vec<I32F32> = vec![
+        I32F32::from_num(-1.0),
+        I32F32::from_num(10.0),
+        I32F32::from_num(30.0),
+    ];
+    inplace_normalize(&mut x2);
+    assert_vec_compare(
+        &x2,
+        &[
+            I32F32::from_num(-0.0256410255),
+            I32F32::from_num(0.2564102563),
+            I32F32::from_num(0.769230769),
+        ],
+        epsilon,
+    );
+}
+
+#[test]
+fn test_math_inplace_normalize_64() {
+    let epsilon: I64F64 = I64F64::from_num(0.0001);
+    let mut x1: Vec<I64F64> = vec![
+        I64F64::from_num(1.0),
+        I64F64::from_num(10.0),
+        I64F64::from_num(30.0),
+    ];
+    inplace_normalize_64(&mut x1);
+    assert_vec_compare_64(
+        &x1,
+        &[
+            I64F64::from_num(0.0243902437),
+            I64F64::from_num(0.243902439),
+            I64F64::from_num(0.7317073171),
+        ],
+        epsilon,
+    );
+    let mut x2: Vec<I64F64> = vec![
+        I64F64::from_num(-1.0),
+        I64F64::from_num(10.0),
+        I64F64::from_num(30.0),
+    ];
+    inplace_normalize_64(&mut x2);
+    assert_vec_compare_64(
+        &x2,
+        &[
+            I64F64::from_num(-0.0256410255),
+            I64F64::from_num(0.2564102563),
+            I64F64::from_num(0.769230769),
+        ],
+        epsilon,
+    );
+}
+
+#[test]
+fn test_math_vecdiv() {
+    let x: Vec<I32F32> = vec_to_fixed(&[]);
+    let y: Vec<I32F32> = vec_to_fixed(&[]);
+    let result: Vec<I32F32> = vec_to_fixed(&[]);
+    assert_eq!(result, vecdiv(&x, &y));
+
+    let x: Vec<I32F32> = vec_to_fixed(&[0., 1., 0., 1.]);
+    let y: Vec<I32F32> = vec_to_fixed(&[0., 1., 1., 0.]);
+    let result: Vec<I32F32> = vec_to_fixed(&[0., 1., 0., 0.]);
+    assert_eq!(result, vecdiv(&x, &y));
+
+    let x: Vec<I32F32> = vec_to_fixed(&[1., 1., 10.]);
+    let y: Vec<I32F32> = vec_to_fixed(&[2., 3., 2.]);
+    let result: Vec<I32F32> = vec![fixed(1.) / fixed(2.), fixed(1.) / fixed(3.), fixed(5.)];
+    assert_eq!(result, vecdiv(&x, &y));
+}
+
+#[test]
+fn test_math_inplace_row_normalize() {
+    let epsilon: I32F32 = I32F32::from_num(0.0001);
+    let vector: Vec<f32> = vec![
+        0., 1., 2., 3., 4., 0., 10., 100., 1000., 10000., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1.,
+    ];
+    let mut mat = vec_to_mat_fixed(&vector, 4, false);
+    inplace_row_normalize(&mut mat);
+    let target: Vec<f32> = vec![
+        0., 0.1, 0.2, 0.3, 0.4, 0., 0.0009, 0.009, 0.09, 0.9, 0., 0., 0., 0., 0., 0.2, 0.2, 0.2,
+        0.2, 0.2,
+    ];
+    assert_mat_compare(&mat, &vec_to_mat_fixed(&target, 4, false), epsilon);
+}
+
```

Comment:
Perhaps a check that `normalize` handles cases where the input vector sum is zero to avoid division by zero?

```suggestion

#[test]
fn test_math_normalize_zero_sum() {
    let epsilon: I32F32 = I32F32::from_num(0.0001);
    let x: Vec<I32F32> = vec![I32F32::from_num(0.0), I32F32::from_num(0.0)];
    let y: Vec<I32F32> = normalize(&x);
    assert_vec_compare(&x, &y, epsilon); // Expect y to handle zero-sum case gracefully
}


```

Reply:
> Perhaps a check that `normalize` handles cases where the input vector sum is zero to avoid division by zero?

also an existing test, that was not added in this pr